hazelcast/src/main/java/com/hazelcast/sql/impl/expression/datetime/ToTimestampTzFunction.java

```diff
@@ -0,0 +1,92 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.datetime;
+
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.UniExpressionWithType;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalUnit;
+
+public final class ToTimestampTzFunction extends UniExpressionWithType<OffsetDateTime> implements IdentifiedDataSerializable {
+    private static final long YEAR_MILLISECONDS = 31536000000L;
+    private static final long YEAR_MICROSECONDS = YEAR_MILLISECONDS * 1000L;
+    private static final long YEAR_NANOSECONDS = YEAR_MICROSECONDS * 1000L;
+
+    public ToTimestampTzFunction() { }
+
+    private ToTimestampTzFunction(Expression<?> operand, QueryDataType resultType) {
+        super(operand, resultType);
+    }
+
+    public static ToTimestampTzFunction create(Expression<?> operand, QueryDataType resultType) {
+        return new ToTimestampTzFunction(operand, resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_TO_TIMESTAMP_TZ;
+    }
+
+    @Override
+    public OffsetDateTime eval(final Row row, final ExpressionEvalContext context) {
+        final Number value = (Number) this.operand.eval(row, context);
+        if (value == null) {
+            return null;
+        }
+
+        final long unixTimestamp = value.longValue();
+        final TemporalUnit unit = getChronoUnit(unixTimestamp);
+        final Instant instant = Instant.EPOCH.plus(unixTimestamp, unit);
+
+        return OffsetDateTime.from(instant.atZone(ZoneOffset.systemDefault()));
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+    }
+
+    private TemporalUnit getChronoUnit(final long unixTimestamp) {
+        // -Long.MIN_VALUE returns Long.MIN_VALUE, probably because an absolute value of that is Long.MAX_VALUE + 1.
+        final long absTimestamp = unixTimestamp == Long.MIN_VALUE
```

Comment:
Actually this is no longer the Snowflake semantics. And I think their semantics makes more sense. If you work with epoch-seconds, every value from -999999 to 9999 will work for you. If you work with epoch-millis or a finer unit, values >=1971-01-01T00:00 UTC will work. With ABS, you can't use values before 1970 at all.

Reply:
So this `ABS` is here only to judge which unit is to be used with the actual computation (L67), so this should still allow to work with negative timestamps.  

I think their semantics to treat any negative timestamp as in seconds is more of a placeholder/temporary solution (judging by that caution section in docs - https://docs.snowflake.com/en/sql-reference/functions/to_timestamp.html#usage-notes). I think it would be worth keeping this logic here, wdyt? 

Maybe a better solution is to provide a second optional argument for this function which would allow overriding the timeunit?

Reply:
As I explained, I think it's actually better the way they have it now, not using ABS. For epoch-millis, the usable range is after 1971. It seems weird to have some millisecond support before 1969. Same for micros and nanos. 
When you're working with epoch-seconds, with ABS your year range is actually ~970..2970. Without ABS, the range is "unlimited" from below...

Practically it's the same, but from theoretical point of view, not using ABS makes more sense. And we'll at least be compatible with someone :)